aten/src/ATen/native/cpu/Depthwise.cpp

```diff
@@ -0,0 +1,296 @@
+#include <arm_neon.h>
+#endif
+
+namespace at {
+namespace native {
+namespace {
+
+struct DepthwiseArgs final {
+  // Input layer dimensions
+  int batch;
+  int in_rows;
+  int in_cols;
+  int stride;
+  int pad_rows;
+  int pad_cols;
+
+  // Output layer dimensions
+  int out_rows;
+  int out_cols;
+};
+
+inline std::vector<int64_t> calculate_conv_output_size(
+    const IntArrayRef input_size,
+    const IntArrayRef weight_size,
+    const IntArrayRef padding,
+    const IntArrayRef stride) {
+  const auto calc_output_dimension = [](
+    const int input, const int kernel, const int padding, const int stride) {
+    return 1 + (input - kernel + 2 * padding) / stride;
+  };
+
+  return std::vector<int64_t> {
+    input_size[0],
+    weight_size[0],
+    calc_output_dimension(input_size[2], weight_size[2], padding[0], stride[0]),
+    calc_output_dimension(input_size[3], weight_size[3], padding[1], stride[1]),
+  };
+}
+
+#ifdef __ARM_NEON__
+
+inline void winograd_f2k3_input_transform_inplace__neon(
+    float32x4_t* const d0,
+    float32x4_t* const d1,
+    float32x4_t* const d2,
+    float32x4_t* const d3) {
+  const float32x4_t wd0 = *d0 - *d2;
+  const float32x4_t wd1 = *d1 + *d2;
+  const float32x4_t wd2 = -*d1 + *d2;
+  const float32x4_t wd3 = *d1 - *d3;
+  *d0 = wd0;
+  *d1 = wd1;
+  *d2 = wd2;
+  *d3 = wd3;
+}
+
+inline void winograd_f2k3_output_transform_inplace__neon(
+    float32x4_t* const m0,
+    float32x4_t* const m1,
+    const float32x4_t* const m2,
+    const float32x4_t* const m3) {
+  *m0 = *m0 + *m1 + *m2;
+  *m1 = *m1 - *m2 - *m3;
+}
+
+inline float32x4_t
+vmuladdq_f32(const float32x4_t c, const float32x4_t a, const float32x4_t b) {
+#if defined(__aarch64__)
+  return vfmaq_f32(c, a, b);
+#else
+  return vmlaq_f32(c, a, b);
+#endif
+}
+
+inline float32x4_t
+vmulsubq_f32(const float32x4_t c, const float32x4_t a, const float32x4_t b) {
+#if defined(__aarch64__)
+  return vfmsq_f32(c, a, b);
+#else
+  return vmlsq_f32(c, a, b);
+#endif
+}
+
+inline void winograd_f2k3_kernel_transform__neon(
+    const float32x4_t g0,
+    const float32x4_t g1,
+    const float32x4_t g2,
+    float32x4_t* const transform0,
+    float32x4_t* const transform1,
+    float32x4_t* const transform2,
+    float32x4_t* const transform3) {
+  const float32x4_t const_half = vdupq_n_f32(0.5f);
+  float32x4_t half_g0_plus_g2 = const_half * (g0 + g2);
+  *transform0 = g0;
+  *transform1 = vmuladdq_f32(half_g0_plus_g2, const_half, g1);
+  *transform2 = vmulsubq_f32(half_g0_plus_g2, const_half, g1);
+  *transform3 = g2;
+}
+
+inline float32x4x4_t v4f_transpose4x4__neon(const float32x4x4_t m) {
+  float32x4x4_t ret;
+  vst4q_f32((float*)(&ret), m);
+  return ret;
+}
+
+void depthwise3x3ConvImpl(
+    const DepthwiseArgs& args,
+    const float* const input,
+    const float* const kernel,
+    const float* const bias,
+    float* const output) {
+  const float32x4_t vbias = vsetq_lane_f32(*bias, vdupq_n_f32(0.0), 1);
+  float32x4x4_t kernel_tile;
+
+  {
```

Comment:
Do you have a sense of how much time we spend in this block?  If it's significant, we could try to find a prepacking approach that would allow us to cache the results.

Reply:
Good point, I'll follow up with some profiling.